Fix Android openAuth async issues (issue #36)

Luke-Rogerson opened the pull request:
This is to address an issue with the `_openAuthSessionPolyfillAsync` method for Android. See #36 for details.

It ensures that the `_waitForRedirectAsync` method is properly awaited, the InAppBrowser is correctly opened and improves the promise resolution check for the `_redirectHandler`.

viewpnt1 commented:
if an error happens during login flow, _waitForRedirectAsync will never get resolved because no redirect happened server side, or am I wrong?

Luke-Rogerson commented:
@viewpnt1 Sometimes, `_waitForRedirectAsync` is getting immediately resolved as `type: success`, which we don't want, as at that point we don't have the `auth_code` appended to the end of the URL, which is why I added that additional check.

However, perhaps there is something I'm not fully understanding. @jdnichollsc could you help us here and explain a bit more about when `_waitForRedirectAsync` should be resolved? Is the check I added for a param not the correct way to properly wait for a valid response?

Luke-Rogerson commented:
@viewpnt1 Sometimes, `_waitForRedirectAsync` is getting immediately resolved as `type: success`, which we don't want, as at that point we don't have the `auth_code` appended to the end of the URL, which is why I added that additional check.

However, perhaps there is something I'm not fully understanding. @jdnichollsc could you help us here and explain a bit more about when `_waitForRedirectAsync` should be resolved? Is the check I added for a param not the correct way to properly wait for a valid response?

Luke-Rogerson commented on index.js:
```diff
@@ -84,7 +84,10 @@ async function open(
   }
-  return RNInAppBrowser.open(inAppBrowseroptions);
+  // use SetTimeout to make sure InAppBrowser doesn't open until current execution context has finished
+  setTimeout(() => {
```
Do you have any alternatives then? I know it's not a particularly elegant solution, but I was getting very hit or miss results when the in app browser modal was actually opening. Adding the `setTimeout` resolved this.

jdnichollsc replied:
Can you share your code? we don't have any message :)

Luke-Rogerson commented on index.js:
```diff
@@ -130,32 +133,33 @@ async function _openAuthSessionPolyfillAsync(
   options: InAppBrowserOptions
 ): Promise<AuthSessionResult> {
+  // reset redirect handler upon new Auth session
+  _redirectHandler = null
```
OK, fair enough

Luke-Rogerson commented on index.js:
```diff
@@ -130,32 +133,33 @@ async function _openAuthSessionPolyfillAsync(
-      if (event.url.startsWith(returnUrl)) {
+    _redirectHandler = (event: RedirectEvent) => {      
+       // check URL has param before resolving (eg. example://auth?code=abcd) 
+          if (event.url.startsWith(returnUrl + '?')) {
```
Is it not? Are we not always expecting the redirect uri + a `code` param as the resolved value from this function?

jdnichollsc replied:
Params are optional 👍

Luke-Rogerson commented on index.js:
```diff
@@ -130,32 +133,33 @@ async function _openAuthSessionPolyfillAsync(
-      open(startUrl, options),
-      _waitForRedirectAsync(returnUrl)
-    ]);
+    await open(startUrl, options)
```
What kind of delay were you thinking? 1000ms?

jdnichollsc replied:
I mean adding a new option to configure that

Luke-Rogerson commented:
> Check my comments and let me know 👍

Thanks for the review! Please see my responses 😎

jdnichollsc commented:
Thanks guys for your help, we're going to use this other PR instead https://github.com/proyecto26/react-native-inappbrowser/pull/103

Luke-Rogerson commented:
No problem. Apologies I wasn't able to update this PR. I've been away from my computer for the last week or so!
Thanks to you for all your work on this library.

jdnichollsc commented:
No problem bro, I want to fix that issue too
